subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java

```diff
@@ -597,6 +597,25 @@ public int compareTo(Project otherProject) {
         return ProjectOrderingUtil.compare(this, otherProject);
     }
 
+    @Override
+    public int hashCode() {
+        return getIdentityPath().hashCode();
+    }
+
+    @Override
+    public boolean equals(Object otherProject) {
+        if (this == otherProject) {
+            return true;
+        }
+        if (otherProject == null) {
+            return false;
+        }
+        if (otherProject instanceof ProjectInternal) {
+            return getIdentityPath().equals(((ProjectInternal) otherProject).getIdentityPath());
```

Comment:
IMHO this should be the correct implementation, asserting equality by project path.

I'd not even check if it's `ProjectInternal` and just cast to it.